Unable to add a fill-pattern to a layer

<!--
Hello! Thanks for contributing. To help us diagnose your problem quickly, please:

 - Include a link to a minimal demonstration of the bug. We recommend using https://jsbin.com
 - Ensure you can reproduce the bug using the latest release.
 - Check the console for relevant errors and warnings
 - Only post to report a bug or request a feature. Direct all other questions to https://stackoverflow.com/questions/tagged/mapbox-gl-js
-->

**mapbox-gl-js version**: 0.18.0

### Steps to Trigger Behavior

Tested in Chrome 50.

 1. create a layer with a fill-color and no fill-pattern
 2. use `map.setPaintProperty()` to set a fill-pattern

### Expected Behavior

The layer should render with the pattern and continue to be responsive to user input and style mutations.

### Actual Behavior

The pattern is set for the layer. An error is logged to the console and the map is no longer interactive.

    Uncaught AssertionError: trying to bind a VAO to a different shader

<img width="512" alt="screen shot 2016-05-05 at 4 56 28 pm" src="https://cloud.githubusercontent.com/assets/302992/15056973/88bf6a5c-12e2-11e6-8238-5abdaa08b457.png">

cc @lucaswoj

--- a/js/render/vertex_array_object.js
+++ b/js/render/vertex_array_object.js
@@ -27,8 +27,15 @@ VertexArrayObject.prototype.bind = function(gl, program, vertexBuffer, elementBu
         util.warnOnce('Not using VertexArrayObject extension.');
     }
 
-    if (!this.boundProgram) {
+    var isFreshBindRequired = !(
+        this.boundProgram &&
+        this.boundProgram === program &&
+        this.boundVertexBuffer === vertexBuffer &&
+        this.boundVertexBuffer2 === vertexBuffer2 &&
+        this.boundElementBuffer === elementBuffer
+    );
 
+    if (isFreshBindRequired) {
         var numPrevAttributes = ext ? 0 : (gl.currentNumAttributes || 0);
         var numNextAttributes = program.numAttributes;
         var i;
@@ -66,13 +73,6 @@ VertexArrayObject.prototype.bind = function(gl, program, vertexBuffer, elementBu
             this.boundVertexBuffer2 = vertexBuffer2;
             this.boundElementBuffer = elementBuffer;
         }
-
-    } else {
-        // verify that bind was called with the same arguments
-        assert(this.boundProgram === program, 'trying to bind a VAO to a different shader');
-        assert(this.boundVertexBuffer === vertexBuffer, 'trying to bind a VAO to a different vertex buffer');
-        assert(this.boundVertexBuffer2 === vertexBuffer2, 'trying to bind a VAO to a different vertex buffer');
-        assert(this.boundElementBuffer === elementBuffer, 'trying to bind a VAO to a different element buffer');
     }
 };
 

--- a/js/render/vertex_array_object.js
+++ b/js/render/vertex_array_object.js
@@ -15,65 +15,74 @@ function VertexArrayObject() {
 
 VertexArrayObject.prototype.bind = function(gl, program, vertexBuffer, elementBuffer, vertexBuffer2) {
 
-    var ext = gl.extVertexArrayObject;
-    if (ext === undefined) {
-        ext = gl.extVertexArrayObject = gl.getExtension("OES_vertex_array_object");
+    if (gl.extVertexArrayObject === undefined) {
+        gl.extVertexArrayObject = gl.getExtension("OES_vertex_array_object");
     }
 
-    if (ext) {
-        if (!this.vao) this.vao = ext.createVertexArrayOES();
-        ext.bindVertexArrayOES(this.vao);
+    if (this.isFreshBindRequired(gl, program, vertexBuffer, elementBuffer, vertexBuffer2)) {
+        this.freshBind(gl, program, vertexBuffer, elementBuffer, vertexBuffer2);
     } else {
-        util.warnOnce('Not using VertexArrayObject extension.');
+        gl.extVertexArrayObject.bindVertexArrayOES(this.vao);
     }
+};
 
-    var isFreshBindRequired = !(
-        this.boundProgram &&
-        this.boundProgram === program &&
-        this.boundVertexBuffer === vertexBuffer &&
-        this.boundVertexBuffer2 === vertexBuffer2 &&
-        this.boundElementBuffer === elementBuffer
+VertexArrayObject.prototype.isFreshBindRequired = function(gl, program, vertexBuffer, elementBuffer, vertexBuffer2) {
+    return (
+        !this.vao ||
+        !this.boundProgram ||
+        this.boundProgram !== program ||
+        this.boundVertexBuffer !== vertexBuffer ||
+        this.boundVertexBuffer2 !== vertexBuffer2 ||
+        this.boundElementBuffer !== elementBuffer
     );
+};
 
-    if (isFreshBindRequired) {
-        var numPrevAttributes = ext ? 0 : (gl.currentNumAttributes || 0);
-        var numNextAttributes = program.numAttributes;
-        var i;
+VertexArrayObject.prototype.freshBind = function(gl, program, vertexBuffer, elementBuffer, vertexBuffer2) {
+    var numPrevAttributes;
+    var numNextAttributes = program.numAttributes;
 
-        // Enable all attributes for the new program.
-        for (i = numPrevAttributes; i < numNextAttributes; i++) {
-            gl.enableVertexAttribArray(i);
-        }
+    if (gl.extVertexArrayObject) {
+        if (this.vao) this.destroy(gl);
+        this.vao = gl.extVertexArrayObject.createVertexArrayOES();
+        gl.extVertexArrayObject.bindVertexArrayOES(this.vao);
+        numPrevAttributes = 0;
 
-        if (!ext) {
-            // Disable all attributes from the previous program that aren't used in
-            // the new program. Note: attribute indices are *not* program specific!
-            // WebGL breaks if you disable attribute 0. http://stackoverflow.com/questions/20305231
-            assert(i > 0);
-            for (i = numNextAttributes; i < numPrevAttributes; i++) {
-                gl.disableVertexAttribArray(i);
-            }
-            gl.currentNumAttributes = numNextAttributes;
-        }
+        // store the arguments so that we can verify them when the vao is bound again
+        this.boundProgram = program;
+        this.boundVertexBuffer = vertexBuffer;
+        this.boundVertexBuffer2 = vertexBuffer2;
+        this.boundElementBuffer = elementBuffer;
 
-        vertexBuffer.bind(gl);
-        vertexBuffer.setVertexAttribPointers(gl, program);
-        if (vertexBuffer2) {
-            vertexBuffer2.bind(gl);
-            vertexBuffer2.setVertexAttribPointers(gl, program);
-        }
-        if (elementBuffer) {
-            elementBuffer.bind(gl);
-        }
+    } else {
+        numPrevAttributes = gl.currentNumAttributes || 0;
+        util.warnOnce('Not using VertexArrayObject extension.');
 
-        if (ext) {
-            // store the arguments so that we can verify them when the vao is bound again
-            this.boundProgram = program;
-            this.boundVertexBuffer = vertexBuffer;
-            this.boundVertexBuffer2 = vertexBuffer2;
-            this.boundElementBuffer = elementBuffer;
+        // Disable all attributes from the previous program that aren't used in
+        // the new program. Note: attribute indices are *not* program specific!
+        for (var i = numNextAttributes; i < numPrevAttributes; i++) {
+            // WebGL breaks if you disable attribute 0.
+            // http://stackoverflow.com/questions/20305231
+            assert(i !== 0);
+            gl.disableVertexAttribArray(i);
         }
     }
+
+    // Enable all attributes for the new program.
+    for (var j = numPrevAttributes; j < numNextAttributes; j++) {
+        gl.enableVertexAttribArray(j);
+    }
+
+    vertexBuffer.bind(gl);
+    vertexBuffer.setVertexAttribPointers(gl, program);
+    if (vertexBuffer2) {
+        vertexBuffer2.bind(gl);
+        vertexBuffer2.setVertexAttribPointers(gl, program);
+    }
+    if (elementBuffer) {
+        elementBuffer.bind(gl);
+    }
+
+    gl.currentNumAttributes = numNextAttributes;
 };
 
 VertexArrayObject.prototype.unbind = function(gl) {

--- a/js/render/vertex_array_object.js
+++ b/js/render/vertex_array_object.js
@@ -19,22 +19,19 @@ VertexArrayObject.prototype.bind = function(gl, program, vertexBuffer, elementBu
         gl.extVertexArrayObject = gl.getExtension("OES_vertex_array_object");
     }
 
-    if (this.isFreshBindRequired(gl, program, vertexBuffer, elementBuffer, vertexBuffer2)) {
-        this.freshBind(gl, program, vertexBuffer, elementBuffer, vertexBuffer2);
-    } else {
-        gl.extVertexArrayObject.bindVertexArrayOES(this.vao);
-    }
-};
-
-VertexArrayObject.prototype.isFreshBindRequired = function(gl, program, vertexBuffer, elementBuffer, vertexBuffer2) {
-    return (
+    var isFreshBindRequired = (
         !this.vao ||
-        !this.boundProgram ||
         this.boundProgram !== program ||
         this.boundVertexBuffer !== vertexBuffer ||
         this.boundVertexBuffer2 !== vertexBuffer2 ||
         this.boundElementBuffer !== elementBuffer
     );
+
+    if (!gl.extVertexArrayObject || isFreshBindRequired) {
+        this.freshBind(gl, program, vertexBuffer, elementBuffer, vertexBuffer2);
+    } else {
+        gl.extVertexArrayObject.bindVertexArrayOES(this.vao);
+    }
 };
 
 VertexArrayObject.prototype.freshBind = function(gl, program, vertexBuffer, elementBuffer, vertexBuffer2) {

--- a/package.json
+++ b/package.json
@@ -46,7 +46,7 @@
     "express": "^4.13.4",
     "gl": "^2.1.5",
     "istanbul": "^0.4.2",
-    "mapbox-gl-test-suite": "mapbox/mapbox-gl-test-suite#f6ab579451c8d929584d586ecd85b9b38d124a87",
+    "mapbox-gl-test-suite": "mapbox/mapbox-gl-test-suite#8e37472e9b5fd2a3a7385ec68e79b052a9e6035c",
     "nyc": "^6.1.1",
     "sinon": "^1.15.4",
     "st": "^1.0.0",

